src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java

```diff
@@ -0,0 +1,352 @@
+package com.wavefront.sdk.entities.histograms;
+
+import com.tdunning.math.stats.AVLTreeDigest;
+import com.tdunning.math.stats.Centroid;
+import com.tdunning.math.stats.TDigest;
+import com.wavefront.sdk.common.Pair;
+
+import java.lang.ref.Reference;
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static java.lang.Double.NaN;
+
+/**
+ * Wavefront implementation of a histogram
+ *
+ * @author Sushant Dewan (sushant@wavefront.com).
+ */
+public class WavefrontHistogramImpl {
+  private final static int ACCURACY = 100;
+
+  /**
+   * If a thread's bin queue has exceeded MAX_BINS number of bins (e.g., the thread has data
+   * that has yet to be reported for more than MAX_BINS number of minutes), delete the oldest bin.
+   * Defaulted to 10 because we can expect the histogram to be reported at least once every
+   * 10 minutes.
+   */
+  private final static int MAX_BINS = 10;
+
+  private final Supplier<Long> clockMillis;
+
+  /**
+   * Global list of thread local histogramBinsList wrapped in WeakReference
+   */
+  private final List<WeakReference<LinkedList<MinuteBin>>> globalHistogramBinsList =
+      new ArrayList<>();
+
+  /**
+   * ThreadLocal histogramBinsList where the initial value set is also added to a
+   * global list of thread local histogramBinsList wrapped in WeakReference
+   */
+  private final ThreadLocal<LinkedList<MinuteBin>> histogramBinsList =
+      ThreadLocal.withInitial(() -> {
+        LinkedList<MinuteBin> sharedBinsInstance = new LinkedList<>();
+        globalHistogramBinsList.add(new WeakReference<>(sharedBinsInstance));
+        return sharedBinsInstance;
+  });
+
+  public WavefrontHistogramImpl() {
+    this(System::currentTimeMillis);
+  }
+
+  public WavefrontHistogramImpl(Supplier<Long> clockMillis) {
+    this.clockMillis = clockMillis;
+  }
+
+  public void update(int value) {
+    update((double) value);
+  }
+
+  public void update(long value) {
+    update((double) value);
+  }
+
+  public void update(double value) {
+    getCurrentBin().distribution.add(value);
+  }
+
+  /**
+   * Bulk-update this histogram with a set of centroids.
+   *
+   * @param means  the centroid values
+   * @param counts the centroid weights/sample counts
+   */
+  public void bulkUpdate(List<Double> means, List<Integer> counts) {
+    if (means != null && counts != null) {
+      int n = Math.min(means.size(), counts.size());
+      MinuteBin currentBin = getCurrentBin();
+      for (int i = 0; i < n; ++i) {
+        currentBin.distribution.add(means.get(i), counts.get(i));
+      }
+    }
+  }
+
+  /**
+   * @return returns the number of values in the distribution.
+   */
+  public long getCount() {
+    return globalHistogramBinsList.stream().map(Reference::get).filter(Objects::nonNull).
```

Comment:
access to globalHistogramBinList is not protected from multiple threads accessing it. Can probably use a Set that's backed by a Sets.fromMap(new ConcurrentHashMap...), have synchronization or CopyOnWriteArrayList